### PR TITLE
ZCS-9728 Migrate nextcloud to Zimbra9

### PR DESCRIPTION
**Pull request by rupalid**
Migrating changes made to support Nextcloud and  oauth social token refresh from Zimbra X to  Zimbra 9

### CODE REVIEW

File: build.xml [hunk 1]
```diff
@@ -18,22 +21,73 @@
...
-    <ivy:install organisation="com.fasterxml.jackson.core" module="jackson-core" revision="2.10.1" settingsRef="dev.settings" from="chain-resolver" to="build-dist" overwrite="true" transitive="true" type="jar"/>
-    <ivy:install organisation="com.fasterxml.jackson.core" module="jackson-databind" revision="2.10.1" settingsRef="dev.settings" from="chain-resolver" to="build-dist" overwrite="true" transitive="true" type="jar"/>
+    <ivy:install organisation="com.fasterxml.jackson.core" module="jackson-annotations" revision="2.9.2" settingsRef="dev.settings" from="chain-resolver" to="build-dist" overwrite="true" transitive="true" type="jar"/>
+    <ivy:install organisation="com.fasterxml.jackson.core" module="jackson-core" revision="2.9.2" settingsRef="dev.settings" from="chain-resolver" to="build-dist" overwrite="true" transitive="true" type="jar"/>
+    <ivy:install organisation="com.fasterxml.jackson.core" module="jackson-databind" revision="2.9.2" settingsRef="dev.settings" from="chain-resolver" to="build-dist" overwrite="true" transitive="true" type="jar"/>
```
Comment (jrjena136) [inline build.xml:82]: any specific reason to downgrade the version.

**Review by jrjena136 — APPROVED** [review]

File: build.xml [hunk 2]
```diff
@@ -18,22 +21,73 @@
...
 
+  <target name="oauth-social-jar" depends="compile"
+       description="Creates the zm-oauth-social.jar file which is used by for oauth functionality">
+     <antcall target="zimbra-jar">
+       <param name="implementation.title" value="Zimbra Oauth Social Client Jar"/>
+       <param name="jar.file" value="${oauth-social.jar.file}"/>
+       <param name="includes"
+         value="com/zimbra/oauth/handlers/impl/**,
+         com/zimbra/oauth/resources/**,
+         com/zimbra/oauth/schedulers/impl/**,
```
Comment (desouzas) [inline build.xml:47]: These weren't brought over (for obvious compatibility reasons), I'm assuming the task ignores them.

File: build.xml (hunk 1 above)
Comment (rupalid) [reply build.xml:82]: Good catch, Fixed

File: src/java/com/zimbra/oauth/token/handlers/impl/NextCloudTokenHandler.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        String indVa [] = values[0].split(":");
+        final String clientId = indVa[0];
+        final String clientSecret = indVa[1];
+        values = LdapOauthHelper.loadConfiguration(acct, Provisioning.A_zimbraOAuthConsumerRedirectUri,
+            client) ;
+        String redirecUri = values[0].substring(0, values[0].lastIndexOf(":"));
+        final String clientRedirectUri = redirecUri;
+
+        values = LdapOauthHelper.loadConfiguration(acct,  Provisioning.A_zimbraOAuthConsumerAPIScope,client);
+        String nextCloudUrl =  values[0].substring(0, values[0].lastIndexOf(":"));
```
[the remark this answers is not in the document]
Comment (rupalid) [reply src/java/com/zimbra/oauth/token/handlers/impl/NextCloudTokenHandler.java:84]: The load configuration does not load the first configured scope but iterates through all and finds the one with the passed in clientname, in this case nextcloud

File: build.xml (hunk 2 above)
Comment (rupalid) [reply build.xml:47]: No sure, will ping you later

File: src/java/com/zimbra/oauth/token/handlers/impl/NextCloudTokenHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (rupalid) [reply src/java/com/zimbra/oauth/token/handlers/impl/NextCloudTokenHandler.java:84]: Fixed

**Review by jrjena136 — APPROVED** [review]

File: build.xml (hunk 2 above)
Comment (rupalid) [reply build.xml:47]: Talked with  Stephan, this port is good for calssic

**Pull request merged**